server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -270,49 +417,42 @@ static Mapping createDynamicUpdate(DocumentParserContext context) {
     }
 
     static void parseObjectOrNested(DocumentParserContext context) throws IOException {
+        context.publishEvent(new DocumentParserListener.Event.ObjectStart(context.parent()));
```

Comment:
This is really nice. Consider adding a wrapper implementing `Closeable` to play it safe with publishing the `ObjectEnd` event before returning. Same with arrays.

Reply:
I love that idea. I kinda forgot you can do that.